library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileInputOperator.java

```diff
@@ -569,9 +601,18 @@ protected void replay(long windowId)
       Map<Integer, Object> recoveryDataPerOperator = windowDataManager.retrieveAllPartitions(windowId);
 
       for (Object recovery : recoveryDataPerOperator.values()) {
-        @SuppressWarnings("unchecked")
-        LinkedList<RecoveryEntry> recoveryData = (LinkedList<RecoveryEntry>)recovery;
 
+        WindowRecoveryEntry windowRecoveryEntry = (WindowRecoveryEntry)recovery;
+        List<RecoveryEntry> recoveryData = windowRecoveryEntry.fileRecoveryInfo;
+
+        if (windowRecoveryEntry.startBatchEmitted) {
+          emitStartBatchControlTuple();
```

Comment:
@bhupeshchawda for every window that is replayed, replay() performs operations which were done only in that window earlier. Hence, replay ensures idempotency here. You can have a look at the tests added for batch.
Let me know if my understanding regarding your concern is wrong.